src/audio_pixel.cpp

```diff
@@ -0,0 +1,270 @@
+#include "audio_pixel.h"
+#define project nullptr
+
+// audio pixel functions 
+
+double linear_interp(double x, double x1, double x2, double y1, double y2) {
+    // edge cases 
+    if ((x2-x1) == 0)
+        return x1;
+    // interpolation in time domain
+    return (((x2 - x) / (x2 - x1)) * y1) + ((x - x1) / (x2 - x1)) * y2;
+}
+
+double closest_val(double val1, double val2, double target) {
+    return abs(target - val1) <= abs(target - val2) ? val1 : val2;
+}
+
+// ************* audio_pixel_block_t ****************
+
+bool audio_pixel_block_t::flush() const {
+    json j;
+    to_json(j);
+    std::string resource_path = GetResourcePath();
+    std::ofstream ofs;
+    ofs.open(resource_path + "/kiwi-block.json");
+
+    if (!ofs.is_open()){
+        return false;
+    }
+    
+    ofs << std::setw(4) << j << std::endl;
+    return true;
+}
+
+void audio_pixel_block_t::to_json(json& j) const {
+    j[std::to_string(m_pix_per_s)] = m_channel_pixels;
+}
+
+void audio_pixel_block_t::update(std::vector<double>& sample_buffer, int num_channels, int sample_rate) {
+    // constructs an audio pixel block using interleaved sample buffer and num channels
+    // clear the current set of audio pixels 
+    m_channel_pixels.clear();
+    int pixels_per_channel = 0;
+
+    // initalize the m_channel_pixels to have empty audio pixel vectors
+    for (int i = 0; i < num_channels; i++) {
+        std::vector<audio_pixel_t> empty_pixel_vec;
+        m_channel_pixels.push_back(empty_pixel_vec);
+    }
+
+    //calcualte the samples per audio pixel and initalize an audio pixel
+    int samples_per_pixel = sample_rate / m_pix_per_s;
+    m_block_size = ceil((sample_buffer.size()/num_channels)/m_pix_per_s);
+
+    for (int channel= 0; channel < num_channels; channel++) {
+        int collected_samples = 0;
+        audio_pixel_t curr_pixel = audio_pixel_t();
+        pixels_per_channel = 0;
+
+        for (int i = 0; i < sample_buffer.size(); i += num_channels) { 
+            double curr_sample = sample_buffer[i];
+
+            // increment collected samples after creating new audio pixel 
+            collected_samples ++;
```

Comment:
isn't `collected_samples` equal to `i` during it's whole lifetime? if so, could we just use `i` instead?

Reply:
i gotcha, the only situation i was concerned about was the case where the number of samples we have is not divisible by samples per pixel, where we would collect less than samples_per_pixel in a single pixel. but you're right i think i should be able to use: samples_per_pixel - (i % samples_per_pixel)